packages/sandbox/src/cloudwatch_logs_monitor.ts

```diff
@@ -0,0 +1,284 @@
+    }
+
+    this.scheduleNextTick(SLEEP_MS);
+  };
+
+  /**
+   * Reads all new log events from a set of CloudWatch Log Groups in parallel
+   */
+  private readNewEvents = async (): Promise<Array<CloudWatchLogEvent>> => {
+    const promises: Array<Promise<Array<CloudWatchLogEvent>>> = [];
+    for (const logGroup of this.allLogGroups) {
+      promises.push(this.readEventsFromLogGroup(logGroup));
+    }
+    return (await Promise.all(promises)).flat();
+  };
+
+  /**
+   * Print out one CloudWatch event using the local printer.
+   */
+  private print = (event: CloudWatchLogEvent): void => {
+    const cloudWatchEventDisplay =
+      this.logGroupEventDisplay[event.logGroupName];
+    if (!cloudWatchEventDisplay) {
+      return;
+    }
+
+    this.printer.print(
+      `[${format.color(
+        cloudWatchEventDisplay.friendlyName,
+        cloudWatchEventDisplay.color
+      )}] ${format.note(
+        event.timestamp.toLocaleTimeString()
+      )} ${event.message.trim()}`
+    );
+  };
+
+  /**
+   * Reads all new log events from a CloudWatch Log Group
+   * starting at either the time the last deployment was triggered or
+   * when the last event was read on the previous tick
+   */
+  private readEventsFromLogGroup = async (
+    cloudWatchLogsToMonitor: LogGroupStreamingCursor
+  ): Promise<Array<CloudWatchLogEvent>> => {
+    const events: CloudWatchLogEvent[] = [];
+
+    // log events from some service are ingested faster than others
+    // so we need to track the start/end time for each log group individually
+    // to make sure that we process all events from each log group.
+    // endTime tracks the latest event received
+    const startTime = cloudWatchLogsToMonitor.startTime ?? this.startTime;
+    let endTime = startTime;
+    try {
+      const response = await this.cloudWatchLogsClient.send(
+        new FilterLogEventsCommand({
+          logGroupName: cloudWatchLogsToMonitor.logGroupName,
+          limit: 100,
+          startTime,
+        })
+      );
+      const filteredEvents = response.events ?? [];
+
+      for (const event of filteredEvents) {
+        if (event.message) {
+          events.push({
+            message: event.message,
+            logGroupName: cloudWatchLogsToMonitor.logGroupName,
+            timestamp: event.timestamp ? new Date(event.timestamp) : new Date(),
+          });
+
+          if (event.timestamp && endTime < event.timestamp) {
+            endTime = event.timestamp;
+          }
+        }
+      }
+      // As long as there are _any_ events in the log group `filterLogEvents` will return a nextToken.
+      // This is true even if these events are before `startTime`. So if we have 100 events and a nextToken
+      // then assume that we have hit the limit and let the user know some messages have been suppressed.
+      // We are essentially showing them a sampling (10000 events printed out is not very useful)
+      if (filteredEvents.length === 100 && response.nextToken) {
```

Comment:
I think that this shouldn't apply when we're streaming to file.

File is a different case and somebody may want to grep it/search it and missing entries will be surprising there.

Reply:
also, for console. should sampling be configurable through command parameters?

Reply:
> also, for console. should sampling be configurable through command parameters?

This feature is exclusively for sandbox, so won't apply for console at all.

> I think that this shouldn't apply when we're streaming to file.

There is a security aspect here as well to prevent overwhelming the machine with too many logs (DoS)

Reply:
> This feature is exclusively for sandbox, so won't apply for console at all.

I meant `console.log` vs `file.write` scenario on local computer.

Reply:
Oh I see, that is not part of the requirements, but can be easily added in the future if needed.

Reply:
I think this is going to be most wanted feature for people who stream logs to local file.
Mind creating GH issue?

Reply:
https://github.com/aws-amplify/amplify-backend/issues/1625